### PR TITLE
[SYCL] Remove `sycl::feature_not_supported` exception

### PR DESCRIPTION
**Pull request by aelovikov-intel**
One of the SYCL 1.2 exception subclasses being removed during the ABI breaking window since these subclasses have been removed in SYCL 2020.

We still want to be able to carry information about backend error code, so `MPIErr` data member in `sycl::exception` remains. However, since none of the standard ctors can accept it (per the specification), I've introduced a new helper `detail::set_pi_error` to set it. It should only be used when propagating an error comming from the PI/UR level and *NOT* for the error conditions originated in the SYCL RT (which is a change from previous implementation but a logical/justifiable one).

### CODE REVIEW

File: sycl/include/sycl/exception.hpp [hunk 1]
```diff
@@ -70,6 +70,8 @@ class __SYCL_EXPORT SYCLCategory : public std::error_category {
 
 // Forward declare to declare as a friend in sycl::excepton.
 __SYCL_EXPORT pi_int32 get_pi_error(const exception &e);
+// TODO: Should it be exported at all?
+__SYCL_EXPORT exception set_pi_error(exception &&e, pi_int32 pi_err);
```
Comment (maarquitos14) [inline sycl/include/sycl/exception.hpp:74]: Is there any problem if it's not exported? I'd rather not export if if there's no reason.

**Review by maarquitos14 — APPROVED** [review]
LGTM, just a little nit.

File: sycl/include/sycl/exception.hpp (hunk 1 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/exception.hpp:74]: That's essentially the same as https://github.com/intel/llvm/pull/14400/files#r1663304397. I'm not yet sure why it's been done the way it is, but I'd prefer to handle that after old SYCL 1.2 exceptions are removed as I don't want to block other PRs that are dependent on this one.

Comment (maarquitos14) [reply sycl/include/sycl/exception.hpp:74]: That's okay, I'm not blocking the PR because of this.

**Pull request merged**